File: en/java-jaxrs/src/main/java/com/example/controller/UserController.java
package com.example.controller;

import jakarta.ws.rs.*;
import jakarta.ws.rs.core.MediaType;
import jakarta.ws.rs.core.Response;
import com.example.model.User;
import com.example.dto.Result;
import com.example.dto.PageResult;
import java.util.List;

@Path("/api/users")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public class UserController {
    
    @GET
    public Response listUsers(
            @QueryParam("keyword") String keyword,
            @QueryParam("status") Integer status,
            @QueryParam("pageNum") @DefaultValue("1") Integer pageNum,
            @QueryParam("pageSize") @DefaultValue("10") Integer pageSize) {
        PageResult<User> result = new PageResult<>();
        return Response.ok(Result.success(result)).build();
    }
    
    @POST
    public Response createUser(CreateUserReq req) {
        User user = new User();
        return Response.ok(Result.success(user)).build();
    }
    
    @GET
    @Path("/{id}")
    public Response getUser(@PathParam("id") Long id) {
        User user = new User();
        return Response.ok(Result.success(user)).build();
    }
    
    @PUT
    @Path("/{id}")
    public Response updateUser(@PathParam("id") Long id, UpdateUserReq req) {
        User user = new User();
        return Response.ok(Result.success(user)).build();
    }
    
    @DELETE
    @Path("/{id}")
    public Response deleteUser(@PathParam("id") Long id) {
        return Response.ok(Result.success(null)).build();
    }
    
    @POST
    @Path("/{id}/activate")
    public Response activateUser(@PathParam("id") Long id) {
        User user = new User();
        return Response.ok(Result.success(user)).build();
    }
    
    @POST
    @Path("/{id}/deactivate")
    public Response deactivateUser(@PathParam("id") Long id) {
        User user = new User();
        return Response.ok(Result.success(user)).build();
    }
    
    @PUT
    @Path("/{id}/password")
    public Response updatePassword(@PathParam("id") Long id, UpdatePasswordReq req) {
        return Response.ok(Result.success(null)).build();
    }
    
    @PUT
    @Path("/{id}/role")
    public Response updateRole(@PathParam("id") Long id, UpdateRoleReq req) {
        User user = new User();
        return Response.ok(Result.success(user)).build();
    }
    
    @GET
    @Path("/{id}/orders")
    public Response getUserOrders(
            @PathParam("id") Long id,
            @QueryParam("status") Integer status,
            @QueryParam("pageNum") @DefaultValue("1") Integer pageNum,
            @QueryParam("pageSize") @DefaultValue("10") Integer pageSize) {
        PageResult<Object> result = new PageResult<>();
        return Response.ok(Result.success(result)).build();
    }
    
    @GET
    @Path("/{id}/addresses")
    public Response getUserAddresses(@PathParam("id") Long id) {
        List<AddressVO> addresses = List.of();
        return Response.ok(Result.success(addresses)).build();
    }
    
    @POST
    @Path("/{id}/addresses")
    public Response addAddress(@PathParam("id") Long id, CreateAddressReq req) {
        AddressVO address = new AddressVO();
        return Response.ok(Result.success(address)).build();
    }
    
    @PUT
    @Path("/{id}/addresses/{addressId}")
    public Response updateAddress(
            @PathParam("id") Long id,
            @PathParam("addressId") Long addressId,
            UpdateAddressReq req) {
        AddressVO address = new AddressVO();
        return Response.ok(Result.success(address)).build();
    }
    
    @DELETE
    @Path("/{id}/addresses/{addressId}")
    public Response deleteAddress(
            @PathParam("id") Long id,
            @PathParam("addressId") Long addressId) {
        return Response.ok(Result.success(null)).build();
    }
    
    @PUT
    @Path("/{id}/addresses/{addressId}/default")
    public Response setDefaultAddress(
            @PathParam("id") Long id,
            @PathParam("addressId") Long addressId) {
        return Response.ok(Result.success(null)).build();
    }
    
    @GET
    @Path("/check-username")
    public Response checkUsername(@QueryParam("username") String username) {
        Boolean available = true;
        return Response.ok(Result.success(available)).build();
    }
    
    @GET
    @Path("/check-email")
    public Response checkEmail(@QueryParam("email") String email) {
        Boolean available = true;
        return Response.ok(Result.success(available)).build();
    }
    
    @GET
    @Path("/{id}/profile")
    public Response getProfile(@PathParam("id") Long id) {
        UserProfileVO profile = new UserProfileVO();
        return Response.ok(Result.success(profile)).build();
    }
    
    @PUT
    @Path("/{id}/profile")
    public Response updateProfile(@PathParam("id") Long id, UpdateProfileReq req) {
        UserProfileVO profile = new UserProfileVO();
        return Response.ok(Result.success(profile)).build();
    }
    
    @POST
    @Path("/{id}/avatar")
    @Consumes(MediaType.MULTIPART_FORM_DATA)
    public Response uploadAvatar(@PathParam("id") Long id) {
        String avatarUrl = "";
        return Response.ok(Result.success(avatarUrl)).build();
    }
    
    @GET
    @Path("/{id}/wishlist")
    public Response getWishlist(
            @PathParam("id") Long id,
            @QueryParam("pageNum") @DefaultValue("1") Integer pageNum,
            @QueryParam("pageSize") @DefaultValue("10") Integer pageSize) {
        PageResult<Object> result = new PageResult<>();
        return Response.ok(Result.success(result)).build();
    }
    
    @POST
    @Path("/{id}/wishlist/{productId}")
    public Response addToWishlist(
            @PathParam("id") Long id,
            @PathParam("productId") Long productId) {
        return Response.ok(Result.success(null)).build();
    }
    
    @DELETE
    @Path("/{id}/wishlist/{productId}")
    public Response removeFromWishlist(
            @PathParam("id") Long id,
            @PathParam("productId") Long productId) {
        return Response.ok(Result.success(null)).build();
    }
    
    @GET
    @Path("/{id}/statistics")
    public Response getUserStatistics(@PathParam("id") Long id) {
        UserStatisticsVO statistics = new UserStatisticsVO();
        return Response.ok(Result.success(statistics)).build();
    }
    
    @POST
    @Path("/batch")
    public Response batchCreate(List<CreateUserReq> reqs) {
        List<User> users = List.of();
        return Response.ok(Result.success(users)).build();
    }
    
    @DELETE
    @Path("/batch")
    public Response batchDelete(List<Long> ids) {
        return Response.ok(Result.success(null)).build();
    }
    
    @GET
    @Path("/export")
    public Response exportUsers(
            @QueryParam("status") Integer status,
            @QueryParam("startDate") String startDate,
            @QueryParam("endDate") String endDate) {
        String fileUrl = "";
        return Response.ok(Result.success(fileUrl)).build();
    }
    
    @POST
    @Path("/{id}/reset-password")
    public Response resetPassword(@PathParam("id") Long id) {
        return Response.ok(Result.success(null)).build();
    }
    
    @GET
    @Path("/search")
    public Response searchUsers(
            @QueryParam("keyword") String keyword,
            @QueryParam("pageNum") @DefaultValue("1") Integer pageNum,
            @QueryParam("pageSize") @DefaultValue("10") Integer pageSize) {
        PageResult<User> result = new PageResult<>();
        return Response.ok(Result.success(result)).build();
    }
}

class CreateUserReq {
    private String username;
    private String password;
    private String email;
    private String phone;
    private String nickname;
    
    public String getUsername() { return username; }
    public void setUsername(String username) { this.username = username; }
    public String getPassword() { return password; }
    public void setPassword(String password) { this.password = password; }
    public String getEmail() { return email; }
    public void setEmail(String email) { this.email = email; }
    public String getPhone() { return phone; }
    public void setPhone(String phone) { this.phone = phone; }
    public String getNickname() { return nickname; }
    public void setNickname(String nickname) { this.nickname = nickname; }
}

class UpdateUserReq {
    private String nickname;
    private String email;
    private String phone;
    private String avatar;
    
    public String getNickname() { return nickname; }
    public void setNickname(String nickname) { this.nickname = nickname; }
    public String getEmail() { return email; }
    public void setEmail(String email) { this.email = email; }
    public String getPhone() { return phone; }
    public void setPhone(String phone) { this.phone = phone; }
    public String getAvatar() { return avatar; }
    public void setAvatar(String avatar) { this.avatar = avatar; }
}

class UpdatePasswordReq {
    private String oldPassword;
    private String newPassword;
    
    public String getOldPassword() { return oldPassword; }
    public void setOldPassword(String oldPassword) { this.oldPassword = oldPassword; }
    public String getNewPassword() { return newPassword; }
    public void setNewPassword(String newPassword) { this.newPassword = newPassword; }
}

class UpdateRoleReq {
    private String role;
    
    public String getRole() { return role; }
    public void setRole(String role) { this.role = role; }
}

class AddressVO {
    private Long id;
    private String receiverName;
    private String phone;
    private String province;
    private String city;
    private String district;
    private String detailAddress;
    private Boolean isDefault;
    
    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public String getReceiverName() { return receiverName; }
    public void setReceiverName(String receiverName) { this.receiverName = receiverName; }
    public String getPhone() { return phone; }
    public void setPhone(String phone) { this.phone = phone; }
    public String getProvince() { return province; }
    public void setProvince(String province) { this.province = province; }
    public String getCity() { return city; }
    public void setCity(String city) { this.city = city; }
    public String getDistrict() { return district; }
    public void setDistrict(String district) { this.district = district; }
    public String getDetailAddress() { return detailAddress; }
    public void setDetailAddress(String detailAddress) { this.detailAddress = detailAddress; }
    public Boolean getIsDefault() { return isDefault; }
    public void setIsDefault(Boolean isDefault) { this.isDefault = isDefault; }
}

class CreateAddressReq {
    private String receiverName;
    private String phone;
    private String province;
    private String city;
    private String district;
    private String detailAddress;
    
    public String getReceiverName() { return receiverName; }
    public void setReceiverName(String receiverName) { this.receiverName = receiverName; }
    public String getPhone() { return phone; }
    public void setPhone(String phone) { this.phone = phone; }
    public String getProvince() { return province; }
    public void setProvince(String province) { this.province = province; }
    public String getCity() { return city; }
    public void setCity(String city) { this.city = city; }
    public String getDistrict() { return district; }
    public void setDistrict(String district) { this.district = district; }
    public String getDetailAddress() { return detailAddress; }
    public void setDetailAddress(String detailAddress) { this.detailAddress = detailAddress; }
}

class UpdateAddressReq {
    private String receiverName;
    private String phone;
    private String province;
    private String city;
    private String district;
    private String detailAddress;
    
    public String getReceiverName() { return receiverName; }
    public void setReceiverName(String receiverName) { this.receiverName = receiverName; }
    public String getPhone() { return phone; }
    public void setPhone(String phone) { this.phone = phone; }
    public String getProvince() { return province; }
    public void setProvince(String province) { this.province = province; }
    public String getCity() { return city; }
    public void setCity(String city) { this.city = city; }
    public String getDistrict() { return district; }
    public void setDistrict(String district) { this.district = district; }
    public String getDetailAddress() { return detailAddress; }
    public void setDetailAddress(String detailAddress) { this.detailAddress = detailAddress; }
}

class UserProfileVO {
    private Long id;
    private String username;
    private String nickname;
    private String email;
    private String phone;
    private String avatar;
    private String bio;
    private String gender;
    private String birthday;
    
    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public String getUsername() { return username; }
    public void setUsername(String username) { this.username = username; }
    public String getNickname() { return nickname; }
    public void setNickname(String nickname) { this.nickname = nickname; }
    public String getEmail() { return email; }
    public void setEmail(String email) { this.email = email; }
    public String getPhone() { return phone; }
    public void setPhone(String phone) { this.phone = phone; }
    public String getAvatar() { return avatar; }
    public void setAvatar(String avatar) { this.avatar = avatar; }
    public String getBio() { return bio; }
    public void setBio(String bio) { this.bio = bio; }
    public String getGender() { return gender; }
    public void setGender(String gender) { this.gender = gender; }
    public String getBirthday() { return birthday; }
    public void setBirthday(String birthday) { this.birthday = birthday; }
}

class UpdateProfileReq {
    private String nickname;
    private String bio;
    private String gender;
    private String birthday;
    
    public String getNickname() { return nickname; }
    public void setNickname(String nickname) { this.nickname = nickname; }
    public String getBio() { return bio; }
    public void setBio(String bio) { this.bio = bio; }
    public String getGender() { return gender; }
    public void setGender(String gender) { this.gender = gender; }
    public String getBirthday() { return birthday; }
    public void setBirthday(String birthday) { this.birthday = birthday; }
}

class UserStatisticsVO {
    private Long userId;
    private Integer totalOrders;
    private Integer completedOrders;
    private Integer pendingOrders;
    private Integer totalReviews;
    private Integer wishlistCount;
    
    public Long getUserId() { return userId; }
    public void setUserId(Long userId) { this.userId = userId; }
    public Integer getTotalOrders() { return totalOrders; }
    public void setTotalOrders(Integer totalOrders) { this.totalOrders = totalOrders; }
    public Integer getCompletedOrders() { return completedOrders; }
    public void setCompletedOrders(Integer completedOrders) { this.completedOrders = completedOrders; }
    public Integer getPendingOrders() { return pendingOrders; }
    public void setPendingOrders(Integer pendingOrders) { this.pendingOrders = pendingOrders; }
    public Integer getTotalReviews() { return totalReviews; }
    public void setTotalReviews(Integer totalReviews) { this.totalReviews = totalReviews; }
    public Integer getWishlistCount() { return wishlistCount; }
    public void setWishlistCount(Integer wishlistCount) { this.wishlistCount = wishlistCount; }
}

File: en/kotlin-springmvc/src/main/kotlin/com/example/controller/UserController.kt
package com.example.controller

import com.example.dto.PageResult
import com.example.dto.Result
import com.example.model.User
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/users")
class UserController {
    
    @GetMapping
    fun listUsers(
        @RequestParam(required = false) keyword: String?,
        @RequestParam(required = false) status: Int?,
        @RequestParam(defaultValue = "1") pageNum: Int,
        @RequestParam(defaultValue = "10") pageSize: Int
    ): Result<PageResult<User>> {
        val result = PageResult<User>()
        return Result.success(result)
    }
    
    @PostMapping
    fun createUser(@RequestBody req: CreateUserReq): Result<User> {
        val user = User()
        return Result.success(user)
    }
    
    @GetMapping("/{id}")
    fun getUser(@PathVariable id: Long): Result<User> {
        val user = User()
        return Result.success(user)
    }
    
    @PutMapping("/{id}")
    fun updateUser(@PathVariable id: Long, @RequestBody req: UpdateUserReq): Result<User> {
        val user = User()
        return Result.success(user)
    }
    
    @DeleteMapping("/{id}")
    fun deleteUser(@PathVariable id: Long): Result<Void> {
        return Result.success(null)
    }
    
    @PostMapping("/{id}/activate")
    fun activateUser(@PathVariable id: Long): Result<User> {
        val user = User()
        return Result.success(user)
    }
    
    @PostMapping("/{id}/deactivate")
    fun deactivateUser(@PathVariable id: Long): Result<User> {
        val user = User()
        return Result.success(user)
    }
    
    @PutMapping("/{id}/password")
    fun updatePassword(@PathVariable id: Long, @RequestBody req: UpdatePasswordReq): Result<Void> {
        return Result.success(null)
    }
    
    @PutMapping("/{id}/role")
    fun updateRole(@PathVariable id: Long, @RequestBody req: UpdateRoleReq): Result<User> {
        val user = User()
        return Result.success(user)
    }
    
    @GetMapping("/{id}/orders")
    fun getUserOrders(
        @PathVariable id: Long,
        @RequestParam(required = false) status: Int?,
        @RequestParam(defaultValue = "1") pageNum: Int,
        @RequestParam(defaultValue = "10") pageSize: Int
    ): Result<PageResult<Any>> {
        val result = PageResult<Any>()
        return Result.success(result)
    }
    
    @GetMapping("/{id}/addresses")
    fun getUserAddresses(@PathVariable id: Long): Result<List<AddressVO>> {
        val addresses = listOf<AddressVO>()
        return Result.success(addresses)
    }
    
    @PostMapping("/{id}/addresses")
    fun addAddress(@PathVariable id: Long, @RequestBody req: CreateAddressReq): Result<AddressVO> {
        val address = AddressVO()
        return Result.success(address)
    }
    
    @PutMapping("/{id}/addresses/{addressId}")
    fun updateAddress(
        @PathVariable id: Long,
        @PathVariable addressId: Long,
        @RequestBody req: UpdateAddressReq
    ): Result<AddressVO> {
        val address = AddressVO()
        return Result.success(address)
    }
    
    @DeleteMapping("/{id}/addresses/{addressId}")
    fun deleteAddress(@PathVariable id: Long, @PathVariable addressId: Long): Result<Void> {
        return Result.success(null)
    }
    
    @PutMapping("/{id}/addresses/{addressId}/default")
    fun setDefaultAddress(@PathVariable id: Long, @PathVariable addressId: Long): Result<Void> {
        return Result.success(null)
    }
    
    @GetMapping("/check-username")
    fun checkUsername(@RequestParam username: String): Result<Boolean> {
        val available = true
        return Result.success(available)
    }
    
    @GetMapping("/check-email")
    fun checkEmail(@RequestParam email: String): Result<Boolean> {
        val available = true
        return Result.success(available)
    }
    
    @GetMapping("/{id}/profile")
    fun getProfile(@PathVariable id: Long): Result<UserProfileVO> {
        val profile = UserProfileVO()
        return Result.success(profile)
    }
    
    @PutMapping("/{id}/profile")
    fun updateProfile(@PathVariable id: Long, @RequestBody req: UpdateProfileReq): Result<UserProfileVO> {
        val profile = UserProfileVO()
        return Result.success(profile)
    }
    
    @PostMapping("/{id}/avatar")
    fun uploadAvatar(@PathVariable id: Long): Result<String> {
        val avatarUrl = ""
        return Result.success(avatarUrl)
    }
    
    @GetMapping("/{id}/wishlist")
    fun getWishlist(
        @PathVariable id: Long,
        @RequestParam(defaultValue = "1") pageNum: Int,
        @RequestParam(defaultValue = "10") pageSize: Int
    ): Result<PageResult<Any>> {
        val result = PageResult<Any>()
        return Result.success(result)
    }
    
    @PostMapping("/{id}/wishlist/{productId}")
    fun addToWishlist(@PathVariable id: Long, @PathVariable productId: Long): Result<Void> {
        return Result.success(null)
    }
    
    @DeleteMapping("/{id}/wishlist/{productId}")
    fun removeFromWishlist(@PathVariable id: Long, @PathVariable productId: Long): Result<Void> {
        return Result.success(null)
    }
    
    @GetMapping("/{id}/statistics")
    fun getUserStatistics(@PathVariable id: Long): Result<UserStatisticsVO> {
        val statistics = UserStatisticsVO()
        return Result.success(statistics)
    }
    
    @PostMapping("/batch")
    fun batchCreate(@RequestBody reqs: List<CreateUserReq>): Result<List<User>> {
        val users = listOf<User>()
        return Result.success(users)
    }
    
    @DeleteMapping("/batch")
    fun batchDelete(@RequestBody ids: List<Long>): Result<Void> {
        return Result.success(null)
    }
    
    @GetMapping("/export")
    fun exportUsers(
        @RequestParam(required = false) status: Int?,
        @RequestParam(required = false) startDate: String?,
        @RequestParam(required = false) endDate: String?
    ): Result<String> {
        val fileUrl = ""
        return Result.success(fileUrl)
    }
    
    @PostMapping("/{id}/reset-password")
    fun resetPassword(@PathVariable id: Long): Result<Void> {
        return Result.success(null)
    }
    
    @GetMapping("/search")
    fun searchUsers(
        @RequestParam keyword: String,
        @RequestParam(defaultValue = "1") pageNum: Int,
        @RequestParam(defaultValue = "10") pageSize: Int
    ): Result<PageResult<User>> {
        val result = PageResult<User>()
        return Result.success(result)
    }
}

data class CreateUserReq(
    val username: String?,
    val password: String?,
    val email: String?,
    val phone: String?,
    val nickname: String?
)

data class UpdateUserReq(
    val nickname: String?,
    val email: String?,
    val phone: String?,
    val avatar: String?
)

data class UpdatePasswordReq(
    val oldPassword: String?,
    val newPassword: String?
)

data class UpdateRoleReq(
    val role: String?
)

data class AddressVO(
    val id: Long? = null,
    val receiverName: String? = null,
    val phone: String? = null,
    val province: String? = null,
    val city: String? = null,
    val district: String? = null,
    val detailAddress: String? = null,
    val isDefault: Boolean? = null
)

data class CreateAddressReq(
    val receiverName: String?,
    val phone: String?,
    val province: String?,
    val city: String?,
    val district: String?,
    val detailAddress: String?
)

data class UpdateAddressReq(
    val receiverName: String?,
    val phone: String?,
    val province: String?,
    val city: String?,
    val district: String?,
    val detailAddress: String?
)

data class UserProfileVO(
    val id: Long? = null,
    val username: String? = null,
    val nickname: String? = null,
    val email: String? = null,
    val phone: String? = null,
    val avatar: String? = null,
    val bio: String? = null,
    val gender: String? = null,
    val birthday: String? = null
)

data class UpdateProfileReq(
    val nickname: String?,
    val bio: String?,
    val gender: String?,
    val birthday: String?
)

data class UserStatisticsVO(
    val userId: Long? = null,
    val totalOrders: Int? = null,
    val completedOrders: Int? = null,
    val pendingOrders: Int? = null,
    val totalReviews: Int? = null,
    val wishlistCount: Int? = null
)
